core/Pos/Core/Slotting.hs

```diff
@@ -15,18 +15,20 @@ import           Universum
 import           Pos.Core.Class     (HasEpochOrSlot (..), getEpochOrSlot)
 import           Pos.Core.Constants (epochSlots, slotSecurityParam)
 import           Pos.Core.Types     (EpochIndex (..), EpochOrSlot (..), FlatSlotId,
-                                     LocalSlotIndex, SlotId (..), epochOrSlot,
+                                     LocalSlotIndex, SlotAmount, SlotId (..), epochOrSlot,
                                      getSlotIndex, mkLocalSlotIndex)
 import           Pos.Util.Util      (leftToPanic)
 
 -- | Flatten 'SlotId' (which is basically pair of integers) into a single number.
 flattenSlotId :: SlotId -> FlatSlotId
-flattenSlotId SlotId {..} =
-    fromIntegral siEpoch * epochSlots + fromIntegral (getSlotIndex siSlot)
+flattenSlotId SlotId {..} = fromIntegral $
+    (fromIntegral siEpoch * epochSlots :: SlotAmount) +
+    (fromIntegral (getSlotIndex siSlot) :: SlotAmount)
 
 -- | Flattens 'EpochIndex' into a single number.
 flattenEpochIndex :: EpochIndex -> FlatSlotId
-flattenEpochIndex EpochIndex {..} = getEpochIndex * epochSlots
+flattenEpochIndex (EpochIndex i) = fromIntegral $
+    ((fromIntegral i * epochSlots) :: SlotAmount)
```

Comment:
Why do you write `SlotAmount` in expressions which include `epochSlots`? `epochSlots` is `SlotAmount`, so everything should be deduced automatically. Do you think it makes the code more understandable? I am not sure about it, because it's quite obvious that `epochSlots` has type `SlotAmount` as long as the reader knows about existence of such type.